src/audio_pixel.cpp

```diff
@@ -0,0 +1,183 @@
+
+    // perform interpolation
+    audio_pixel_block_t interpolated_block = interpolate_block(t0, t1, pix_per_s, nearest_pps);
+    return interpolated_block;
+};
+
+double audio_pixel_mipmap_t::get_nearest_pps(double pix_per_s) {
+    // TODO: make this a vector of doubles, return the two nearest pps
+    // returning the nearest to first get bilinear interpolation working
+    // 2 edge cases
+    int n_blocks = m_block_pps.size();
+
+    if (pix_per_s <= m_block_pps[0]) // given pps smaller than our smallest rep
+        return m_block_pps[0];
+    else if (pix_per_s >= m_block_pps[n_blocks - 1]) // given pps larger than our largest rep
+        return m_block_pps[n_blocks - 1];
+
+    return get_nearest_block_pps(pix_per_s);
+}
+
+double audio_pixel_mipmap_t::linear_interp(double x, double x1, double x2, double y1, double y2) {
+    // interpolation in time domain
+    return (((x2 - x) / (x2 - x1)) * y1) + ((x - x1) / (x2 - x1)) * y2;
+}
+
+audio_pixel_block_t audio_pixel_mipmap_t::interpolate_block(double t0, double t1, double new_pps, double nearest_pps) {
+    // interpolate from the block with nearest-pps to create a new block at new_pps
+    // the tricky part here is keeping track of the which samples from the nearest block are towards us
+    audio_pixel_block_t nearest_block = m_block_pps[nearest_pps];
+    std::vector<std::vector<audio_pixel_t>> nearest_channel_pixels = nearest_block.get_pixels(t0, t1);
+    int new_number_pixels = (int)round((t1 - t0) * new_pps);
+
+    double new_t_unit = 1/new_pps, nearest_t_unit = 1/nearest_pps; // new/nearest block's time unit (the time between pixels)
+    double nearest_t0 = 0, nearest_t1 = nearest_t_unit; // nearest block's time range and index (position in time of nearest samples)
+    audio_pixel_t nearest_pixel0, nearest_pixel1;
+    int nearest_pixel_idx0 = 0, nearest_pixel_idx1 = 1; 
+    double curr_t = 0; // new block's time 
+
+    std::vector<std::vector<audio_pixel_t>> new_block;
+
+    for (auto& nearest_pixels : nearest_channel_pixels) {
+        // nearest pixels of the specfic channel 
+        audio_pixel_t curr_audio_pixel;
+        std::vector<audio_pixel_t> curr_channel_pixels;
+
+        for (int i = 0; i < new_number_pixels; i++) {
+            curr_t = i * new_t_unit;
+            curr_audio_pixel = audio_pixel_t();
+            
+            // grab the nearest audio pixels and their times
+            nearest_pixel_idx0 = floor(nearest_pps * curr_t);
+            nearest_pixel_idx1 = ceil(nearest_pps * curr_t);
+            nearest_t0 = nearest_pixel_idx0 * nearest_t_unit;
+            nearest_t1 = nearest_t0 + nearest_t_unit;
+            nearest_pixel0 = nearest_pixels[nearest_pixel_idx0];
+            nearest_pixel1 = nearest_pixels[nearest_pixel_idx0];
+
+            // perform liner interpolation for each field of the audio pixel
+            curr_audio_pixel.m_max = linear_interp(curr_t, nearest_t0, nearest_t1, nearest_pixel0.m_max, nearest_pixel0.m_max);
```

Comment:
could we abstract this 3 line chunk to a `linear_interpolation` method in the audio_pixel_t? so this class can remain agnostic to the fields in an `audio_pixel_t`